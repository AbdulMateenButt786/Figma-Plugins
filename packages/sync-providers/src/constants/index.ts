export * from './Providers';
export * from './TokenSetStatus';
export * from './SystemFilenames';
export * from './ErrorMessages';
export * from './TokenTypes';
export * from './ColorModifierTypes';
export * from './ColorSpaceTypes';
export * from './Properties';
