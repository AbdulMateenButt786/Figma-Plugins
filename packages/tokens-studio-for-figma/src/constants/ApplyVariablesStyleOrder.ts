export enum ApplyVariablesStylesOrRawValues {
  VARIABLES_STYLES = 'variables_styles',
  RAW_VALUES = 'raw_values',
}
