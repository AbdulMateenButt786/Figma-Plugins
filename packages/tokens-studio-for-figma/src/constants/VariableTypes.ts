import { TokenTypes } from './TokenTypes';

export const tokenTypesToCreateVariable = [
  TokenTypes.DIMENSION,
  TokenTypes.BORDER_RADIUS,
  TokenTypes.BORDER_WIDTH,
  TokenTypes.SPACING,
  TokenTypes.SIZING,
  TokenTypes.BOOLEAN,
  TokenTypes.COLOR,
  TokenTypes.NUMBER,
  TokenTypes.TEXT,
  TokenTypes.FONT_FAMILIES,
  TokenTypes.FONT_SIZES,
  TokenTypes.FONT_WEIGHTS,
  TokenTypes.LINE_HEIGHTS,
  TokenTypes.LETTER_SPACING,
  TokenTypes.PARAGRAPH_SPACING,
  TokenTypes.PARAGRAPH_INDENT,
  TokenTypes.OPACITY,
];
