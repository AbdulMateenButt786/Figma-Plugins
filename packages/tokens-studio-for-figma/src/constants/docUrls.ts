export const docUrls = {
  root: 'https://docs.tokens.studio/?ref=pf',
};
