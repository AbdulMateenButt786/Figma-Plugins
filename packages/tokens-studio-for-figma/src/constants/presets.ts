import defaultJSON from '@/config/default.json';
import tailwindJSON from '@/config/tailwind.json';
import designTokenStarter from '@/config/designtokensstarter.json';
import { SetTokenDataPayload } from '@/types/payloads';
import { DeepTokensMap } from '@/types';

export type Preset = {
  id: string;
  text: string;
  description: string;
  author: string;
  type: string;
  json: DeepTokensMap;
};

export const presets: Preset[] = [
  {
    id: 'default',
    text: 'Default Preset',
    description: 'A wide variety of tokens and some token sets to give you an idea of what you can do.',
    author: 'Jan Six',
    type: 'default',
    json: defaultJSON,
  },
  {
    id: 'tailwind',
    text: 'Tailwind Preset',
    description: 'A set of tokens that match the Tailwind CSS framework.',
    author: 'Sam Gordashko',
    type: 'tailwind',
    json: tailwindJSON,
  },
  {
    id: 'design-token-starter',
    text: 'Design Token Starter Preset',
    description: 'A set of tokens that match the Design Token Starter.',
    author: 'Philipp Jeroma',
    type: 'design-token-starter',
    json: designTokenStarter,
  },
];
