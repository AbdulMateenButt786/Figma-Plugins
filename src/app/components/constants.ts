export const DEFAULT_DEPTH_LEVEL = 3;
