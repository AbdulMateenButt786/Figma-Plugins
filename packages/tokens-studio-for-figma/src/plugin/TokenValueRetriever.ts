import { ApplyVariablesStylesOrRawValues } from '@/constants/ApplyVariablesStyleOrder';
import { RawVariableReferenceMap } from '@/types/RawVariableReferenceMap';
import { AnyTokenList } from '@/types/tokens';

export class TokenValueRetriever {
  public tokens;

  public variableReferences;

  public cachedVariableReferences;

  private styleReferences;

  private potentialStylePathPrefixes;

  private ignoreFirstPartForStyles;

  public applyVariablesStylesOrRawValue;

  public createStylesWithVariableReferences;

  private getAdjustedTokenName(tokenName: string, internalParent: string | undefined): string {
    const matchingStylePathPrefix = this.potentialStylePathPrefixes?.find((prefix) => prefix === internalParent);

    const withPrefix = [matchingStylePathPrefix, tokenName].filter((n) => n).join('.');

    return withPrefix;
  }

  public initiate({
    tokens,
    variableReferences,
    styleReferences,
    potentialStylePathPrefixes,
    ignoreFirstPartForStyles = false,
    createStylesWithVariableReferences = false,
    applyVariablesStylesOrRawValue = ApplyVariablesStylesOrRawValues.VARIABLES_STYLES,
  }: { tokens: AnyTokenList,
    variableReferences?: RawVariableReferenceMap,
    styleReferences?: Map<string,
    string>,
    potentialStylePathPrefixes?: string[],
    ignoreFirstPartForStyles?: boolean,
    createStylesWithVariableReferences?: boolean,
    applyVariablesStylesOrRawValue?: ApplyVariablesStylesOrRawValues,
  }) {
    this.potentialStylePathPrefixes = typeof potentialStylePathPrefixes !== 'undefined' ? potentialStylePathPrefixes : null;
    this.ignoreFirstPartForStyles = ignoreFirstPartForStyles;
    this.createStylesWithVariableReferences = createStylesWithVariableReferences;
    this.styleReferences = styleReferences || new Map();
    this.variableReferences = variableReferences || new Map();
    this.cachedVariableReferences = new Map();
    this.applyVariablesStylesOrRawValue = applyVariablesStylesOrRawValue;

    this.tokens = new Map<string, any>(tokens.map((token) => {
      const variableId = variableReferences?.get(token.name);
      // For styles, we need to ignore the first part of the token name as well as consider theme prefix
      const adjustedTokenName = this.getAdjustedTokenName(token.name, token.internal__Parent);
      const styleId = styleReferences?.get(adjustedTokenName);
      return [token.name, {
        ...token, variableId, styleId, adjustedTokenName,
      }];
    }));
  }

  public get(tokenName: string) {
    
    console.log("prefixes", this.potentialStylePathPrefixes);
    return this.tokens.get(tokenName);
  }

  public async getVariableReference(tokenName: string) {
    let variable;
    const storedToken = this.tokens.get(tokenName);
    const isUsingReference = storedToken?.rawValue?.startsWith?.('{') && storedToken?.rawValue?.endsWith?.('}');
    const hasCachedVariable = this.cachedVariableReferences.has(tokenName);
    if (hasCachedVariable) {
      variable = this.cachedVariableReferences.get(tokenName);
      return variable;
    }
    const variableMapped = this.variableReferences.get(tokenName) || (isUsingReference ? this.variableReferences.get(storedToken.rawValue.slice(1, -1)) : null);
    if (!variableMapped) return false;
    if (!hasCachedVariable && typeof variableMapped === 'string') {
      try {
        const foundVariable = await figma.variables.importVariableByKeyAsync(variableMapped);
        if (foundVariable) {
          this.cachedVariableReferences.set(tokenName, foundVariable);
          variable = foundVariable;
          return variable;
        }
      } catch (e) {
        console.log('error importing variable', e);
        Promise.reject(e);
      }
    }

    if (variable === undefined) return null;
    return null;
  }

  public getTokens() {
    return this.tokens;
  }

  public clearCache() {
    if (this.cachedVariableReferences) this.cachedVariableReferences.clear();
    if (this.tokens) this.tokens.clear();
    if (this.variableReferences) this.variableReferences.clear();
    if (this.styleReferences) this.styleReferences.clear();
    if (this.potentialStylePathPrefixes) this.potentialStylePathPrefixes = undefined;
    if (this.ignoreFirstPartForStyles) this.ignoreFirstPartForStyles = undefined;
    if (this.createStylesWithVariableReferences) this.createStylesWithVariableReferences = undefined;
  }
}

const defaultTokenValueRetriever = new TokenValueRetriever();

export { defaultTokenValueRetriever };
