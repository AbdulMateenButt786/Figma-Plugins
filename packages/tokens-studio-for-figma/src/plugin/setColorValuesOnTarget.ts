import { isPaintEqual } from '@/utils/isPaintEqual';
import { convertToFigmaColor } from './figmaTransforms/colors';
import { convertStringToFigmaGradient } from './figmaTransforms/gradients';
import { defaultTokenValueRetriever } from './TokenValueRetriever';
import { ColorPaintType, tryApplyColorVariableId } from '@/utils/tryApplyColorVariableId';
import { unbindVariableFromTarget } from './unbindVariableFromTarget';

export default async function setColorValuesOnTarget(target: BaseNode | PaintStyle, token: string, key: 'paints' | 'fills' | 'strokes' = 'paints') {
  const shouldCreateStylesWithVariables = defaultTokenValueRetriever.createStylesWithVariableReferences;
  try {
    const resolvedToken = defaultTokenValueRetriever.get(token);
    if (typeof resolvedToken === 'undefined') throw new Error(`Token ${token} not found in token map`);
    const { description, value } = resolvedToken;
    const resolvedValue = defaultTokenValueRetriever.get(token)?.rawValue;
    if (typeof resolvedValue === 'undefined') throw new Error(`Token ${token} has no resolved value`);
    let existingPaint: Paint | null = null;
    if (key === 'paints' && 'paints' in target) {
      existingPaint = target.paints[0] ?? null;
    } else if (key === 'fills' && 'fills' in target && target.fills !== figma.mixed) {
      existingPaint = target.fills[0] ?? null;
    } else if (key === 'strokes' && 'strokes' in target) {
      existingPaint = target.strokes[0] ?? null;
    }

    if (value.startsWith('linear-gradient')) {
      const { gradientStops, gradientTransform } = convertStringToFigmaGradient(value);
      const newPaint: GradientPaint = {
        type: 'GRADIENT_LINEAR',
        gradientTransform,
        gradientStops,
      };

      if (!existingPaint || !isPaintEqual(newPaint, existingPaint)) {
        if (key === 'paints' && 'paints' in target) target.paints = [newPaint];
        if (key === 'fills' && 'fills' in target) target.fills = [newPaint];
        if (key === 'strokes' && 'strokes' in target) target.strokes = [newPaint];
      }
    } else {
      // If the raw value is a pure reference to another token, we first should try to apply that reference as a variable if it exists.
      let successfullyAppliedVariable = false;
      if (resolvedValue.toString().startsWith('{') && resolvedValue.toString().endsWith('}') && shouldCreateStylesWithVariables) {
        try {
          successfullyAppliedVariable = await tryApplyColorVariableId(target, resolvedValue.slice(1, -1), ColorPaintType.PAINTS);
        } catch (e) {
          console.error('Error setting bound variable for paint', e);
        }
      }
      // If we didnt find a variable to apply, we should apply the color directly.
      if (!successfullyAppliedVariable) {
        const { color, opacity } = convertToFigmaColor(value);
        const newPaint: SolidPaint = { color, opacity, type: 'SOLID' };
        await unbindVariableFromTarget(target, key, newPaint);
        if (!existingPaint || !isPaintEqual(newPaint, existingPaint)) {
          if (key === 'paints' && 'paints' in target) target.paints = [newPaint];
          if (key === 'fills' && 'fills' in target) target.fills = [newPaint];
          if (key === 'strokes' && 'strokes' in target) target.strokes = [newPaint];
        }
      }
    }
    if (description && 'description' in target) {
      target.description = description;
    }
    Promise.resolve();
  } catch (e) {
    // It's not a token value, it's actually a color value
    try {
      const value = token; // the token is not a token, it's a misnomer
      if (value.startsWith('linear-gradient')) {
        const { gradientStops, gradientTransform } = convertStringToFigmaGradient(value);
        const newPaint: GradientPaint = {
          type: 'GRADIENT_LINEAR',
          gradientTransform,
          gradientStops,
        };
        target[key] = [newPaint];
      } else {
        const { color, opacity } = convertToFigmaColor(value);
        const newPaint: SolidPaint = { color, opacity, type: 'SOLID' };
        target[key] = [newPaint];
      }
    } catch (e) {
      console.error('Error setting color', e);
    }
  }
}
