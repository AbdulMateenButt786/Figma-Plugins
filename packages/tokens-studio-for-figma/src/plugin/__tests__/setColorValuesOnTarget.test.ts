import { TokenTypes } from '@/constants/TokenTypes';
import { defaultTokenValueRetriever } from '../TokenValueRetriever';
import setColorValuesOnTarget from '../setColorValuesOnTarget';
import { mockImportVariableByKeyAsync, mockSetBoundVariableForPaint } from '../../../tests/__mocks__/figmaMock';

describe('setColorValuesOnTarget', () => {
  beforeEach(() => {
    defaultTokenValueRetriever.initiate({
      tokens: [{
        name: 'red',
        type: TokenTypes.COLOR,
        value: '#ff0000',
        rawValue: '#ff0000',
        description: 'Red',
      },
      {
        name: 'gradient',
        type: TokenTypes.COLOR,
        value: 'linear-gradient(90deg, #000000 0%, #ffffff 100%)',
        rawValue: 'linear-gradient(90deg, #000000 0%, #ffffff 100%)',
      }, {
        name: 'ref-red',
        type: TokenTypes.COLOR,
        value: '#ff0000',
        rawValue: '{red}',
        description: 'Referenced Red',
      }],
      variableReferences: new Map([['red', '123']]),
      createStylesWithVariableReferences: true,
    });
  });

  it('should be able to update the paints on a style', async () => {
    const mockStyle = {
      type: 'PAINT',
      paints: [],
      description: '',
    } as unknown as PaintStyle;

    await setColorValuesOnTarget(mockStyle, 'red', 'paints');

    expect(mockStyle.description).toEqual('Red');

    expect(mockStyle.paints).toEqual([{
      type: 'SOLID',
      opacity: 1,
      color: { r: 1, g: 0, b: 0 },
    }]);
  });

  it('should be able to update the fills on a node', async () => {
    const mockNode = {
      type: 'RECTANGLE',
      fills: [],
    } as unknown as RectangleNode;

    await setColorValuesOnTarget(mockNode, 'red', 'fills');

    expect(mockNode.fills).toEqual([{
      type: 'SOLID',
      opacity: 1,
      color: { r: 1, g: 0, b: 0 },
    }]);
  });

  it('should be able to update the strokes on a node', async () => {
    const mockNode = {
      type: 'RECTANGLE',
      strokes: [],
    } as unknown as RectangleNode;

    await setColorValuesOnTarget(mockNode, 'red', 'strokes');

    expect(mockNode.strokes).toEqual([{
      type: 'SOLID',
      opacity: 1,
      color: { r: 1, g: 0, b: 0 },
    }]);
  });

  it('should be able to handle a linear gradient', async () => {
    const mockNode = {
      type: 'RECTANGLE',
      fills: [],
    } as unknown as RectangleNode;

    await setColorValuesOnTarget(mockNode, 'gradient', 'fills');

    expect(mockNode.fills).toMatchSnapshot();
  });

  // Test when token doesn't exist
  it('should handle non-existent token gracefully', async () => {
    const mockNode = {
      type: 'RECTANGLE',
      fills: [],
    } as unknown as RectangleNode;

    await setColorValuesOnTarget(mockNode, 'non-existent-token', 'fills');

    // Assert that fills have not been changed
    expect(mockNode.fills).toEqual([]);
  });

  // Test when token value is a reference to another token
  it('should handle token references correctly', async () => {
    const mockStyle = {
      paints: [],
    } as unknown as PaintStyle;

    mockImportVariableByKeyAsync.mockImplementationOnce(() => ({
      id: 'VariableID:3456',
      key: '34567',
      variableCollectionId: 'VariableCollectionId:23:23456',
      name: 'fg/subtle',
    }));

    await setColorValuesOnTarget(mockStyle, 'ref-red', 'paints');

    // Assert that fills have been updated to the referenced token's color
    expect(mockSetBoundVariableForPaint).toHaveBeenCalledWith({
      color: { r: 0, g: 0, b: 0 },
      type: 'SOLID',
    }, 'color', {
      id: 'VariableID:3456', key: '34567', name: 'fg/subtle', variableCollectionId: 'VariableCollectionId:23:23456',
    });
  });

  // Test when token reference is not found
  it('should handle missing token references by applying the hex value', async () => {
    const mockStyle = {
      paints: [],
    } as unknown as PaintStyle;

    mockImportVariableByKeyAsync.mockImplementationOnce(() => null);

    await setColorValuesOnTarget(mockStyle, 'ref-red', 'paints');

    // Assert that fills have been updated to the hex value of the token
    expect(mockSetBoundVariableForPaint).not.toHaveBeenCalledWith();
    expect(mockStyle.paints).toEqual([{
      type: 'SOLID',
      opacity: 1,
      color: { r: 1, g: 0, b: 0 },
    }]);
  });
});
