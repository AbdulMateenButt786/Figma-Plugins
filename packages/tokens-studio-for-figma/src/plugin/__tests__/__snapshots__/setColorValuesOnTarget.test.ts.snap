// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setColorValuesOnTarget should be able to handle a linear gradient 1`] = `
[
  {
    "gradientStops": [
      {
        "color": {
          "a": 1,
          "b": 0,
          "g": 0,
          "r": 0,
        },
        "position": 0,
      },
      {
        "color": {
          "a": 1,
          "b": 1,
          "g": 1,
          "r": 1,
        },
        "position": 1,
      },
    ],
    "gradientTransform": [
      [
        1,
        0,
        0,
      ],
      [
        0,
        1,
        0,
      ],
    ],
    "type": "GRADIENT_LINEAR",
  },
]
`;
